JAPI-318 : provide way to define connect timeout when writing files.

drealeed opened the pull request:
* Create additional constructor for HPCCRemoteFileWriter that takes a connect timeout value
* update current constructor for HPCCRemoteFileWriter to call the new constructor, passing RowServiceOutputWriter.DEFAULT_CONNECT_TIMEOUT_MILLIS
* Create additional constructor for RowServiceOutputWriter that takes a connect timeout value
* update current (non-deprecated) constructor for RowServiceOutputWriter to call the new constructor, passing RowServiceOutputWriter.DEFAULT_CONNECT_TIMEOUT_MILLIS
* update DFSReadWriteTest so that it can run passing a connect timeout value defined as a system property. If no connect timeout is defined, the test runs as it used do with default timeouts specified.

<!-- Thank you for submitting a pull request to the HPCC Java APIs (JAPIS) project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues (https://track.hpccsystems.com/).
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 JAPI-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,and
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and check-off all the items that apply (after pull request has been created)
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [X] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [] I have applied the Eclipse code-format template provided.
- [X] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [X] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [X] I have performed unit tests to cover my changes.
  - [X] I have performed system test and covered possible regressions and side effects.
  - [X] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

## Testing:
<!-- Please describe how this change has been tested.-->
Tested the change by running DFUReadWriteTest with and without the connecttimeoutms system property specified.

 Also, compiled the jar with changes and tested it from the tardis hpcc module. Confirmed that these changes resolved the issues tardis was experiencing with write connection timeouts.

rpastrana commented on dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:
```diff
@@ -342,8 +345,15 @@ private void writeFile(List<HPCCRecord> records, String fileName, FieldDef recor
+                if (connectTimeoutMs != null) {
+                    fileWriter=new HPCCRemoteFileWriter<HPCCRecord>(hpccPartitions[partitionIndex], recordDef,
+                            recordAccessor, CompressionAlgorithm.NONE,connectTimeoutMs);
+                    Thread.sleep(2000);
```
what are we accomplishing with this sleep?
Also, minor but formatting needed.

drealeed replied:
the default timeout ms is 1000. If an override is specified, sleep longer than the default timeout to verify the overridden timeout is actually working.

drealeed replied:
Fixed formatting

rpastrana commented on wsclient/src/main/java/org/hpccsystems/ws/client/platform/test/BaseRemoteTest.java:
```diff
@@ -42,7 +42,7 @@ HPCC SYSTEMS software Copyright (C) 2019 HPCC Systems®.
     protected final static String hpccUser = System.getProperty("hpccuser", defaultUserName);
     protected final static String hpccPass = System.getProperty("hpccpass", "");
-    protected final static String connTO = System.getProperty("connecttimeoutmillis");
+    protected final static Integer connTO = System.getProperty("connecttimeoutmillis")==null?null:Integer.valueOf(System.getProperty("connecttimeoutmillis"));
```
are all references to connTO in the project null safe?
is the ternary expression needed because Integer.valueOf(null) != null?

drealeed replied:
Integer.valueOf(null) throws an error. Yes, the code handles connTO being null. I found one place it wasn't and fixed it and pushed the change to this PR.

rpastrana commented on dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:
```diff
@@ -315,7 +318,7 @@ private void writeFile(List<HPCCRecord> records, String fileName, FieldDef recor
             String clusterName = this.thorcluster;
             System.out.println("Create Start");
-            DFUCreateFileWrapper createResult = dfuClient.createFile(fileName, clusterName, eclRecordDefn, 300, false, DFUFileTypeWrapper.Flat, "");
+            DFUCreateFileWrapper createResult = dfuClient.createFile(fileName, clusterName, eclRecordDefn, connTO, false, DFUFileTypeWrapper.Flat, "");
```
Not sure if important, but this test's connection timeout is different now

drealeed replied:
Fixed this so that if connTO is null the original 300 is used.

rpastrana commented on dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:
```diff
@@ -234,25 +235,27 @@ public void filteredTest() throws Exception
         }
-
+        if (connectTimeoutMillis != null) {
+            file.setFileAccessExpirySecs(connectTimeoutMillis/1000);
```
typically the connect timeout would be considerably shorter than the fileaccessexpiry, i'm not sure what we get from combining the two values. Albeit, we prob do need a junit test w/ very short/log expiry values...

drealeed replied:
I was getting a file access expiry error (connecting over vpn) until I added this to the test.

rpastrana replied:
fair enough, but my issue has to do w/ the reuse of the connTO, we need a dfsclient test specific fileAccessExpiry variable

private final static Integer fileAccessExpiry = System.getProperty("fileAccessExpirySecs")==null?null:Integer.valueOf(System.getProperty("fileAccessExpirySecs"));

rpastrana commented on dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:
```diff
@@ -55,16 +55,16 @@
 {
     private static final String[] datasets       = { "~benchmark::integer::20kb", "~benchmark::all_types::200kb"};
-    private static final int[]    expectedCounts = { 1250, 5600 };
+    private static final int[]    expectedCounts = { 1250, 1558 };
```
was the 5600 record count invalid? @jpmcmu

drealeed replied:
yes, the current file has 1558 recs, I checked on the cluster to confirm.

rpastrana replied:
@drealeed the datasets and expectcounts should really be passed in dynamically, but for now our official build environment's file has 5600 recs. 
let's revert, and we'll provide you the ECL we execute to generate the larger dataset

drealeed commented:
@rpastrana ,all comments and questions fixed/answered.

jpmcmu commented:
@drealeed @rpastrana beyond record count looks good to me. We are generating that dataset in our Jenkins test environment. So the record count is probably different than what is on the 160.

rpastrana commented on dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:
```diff
@@ -55,16 +55,16 @@
-        List<HPCCRecord> records = readFile(file);
-        assertEquals("Not all records loaded",records.size(),5600);
+        List<HPCCRecord> records = readFile(file, connTO);
+        assertEquals("Not all records loaded",expectedCounts[1], records.size());
```
Good catch here, can we also update line 64 to reference datasets[1]?

drealeed commented:
updated file reference to use datasets; reset the file count to 5600; added some logic around the thread.sleep to clarify what it's doing,

Once approved I can squash all these (if I do it now comment pointers will be lost), or you can squash while pulling.

rpastrana commented:
Thanks @drealeed please go ahead and sqaush.

drealeed commented:
@rpastrana squashed
